core/trino-main/src/main/java/io/trino/sql/gen/columnar/DynamicPageFilter.java

```diff
@@ -0,0 +1,196 @@
+{
+    private final Metadata metadata;
+    private final TypeManager typeManager;
+    private final Session session;
+    private final IrExpressionOptimizer irExpressionOptimizer;
+    private final DomainTranslator domainTranslator;
+    private final DynamicFilter dynamicFilter;
+    private final Map<ColumnHandle, Symbol> columnHandles;
+    private final Map<Symbol, Integer> sourceLayout;
+    private final double selectivityThreshold;
+
+    @Nullable
+    @GuardedBy("this")
+    private Supplier<FilterEvaluator> compiledDynamicFilter;
+    @Nullable
+    @GuardedBy("this")
+    private CompletableFuture<?> isBlocked;
+
+    public DynamicPageFilter(
+            PlannerContext plannerContext,
+            Session session,
+            DynamicFilter dynamicFilter,
+            Map<Symbol, ColumnHandle> columnHandles,
+            Map<Symbol, Integer> sourceLayout,
+            double selectivityThreshold)
+    {
+        this.metadata = requireNonNull(plannerContext.getMetadata(), "metadata is null");
+        this.typeManager = requireNonNull(plannerContext.getTypeManager(), "typeManager is null");
+        this.session = requireNonNull(session, "session is null");
+        this.irExpressionOptimizer = newOptimizer(plannerContext);
+        this.domainTranslator = new DomainTranslator(plannerContext.getMetadata());
+        this.dynamicFilter = requireNonNull(dynamicFilter, "dynamicFilter is null");
+        this.columnHandles = columnHandles.entrySet()
+                .stream()
+                .collect(toImmutableMap(Map.Entry::getValue, Map.Entry::getKey));
+        this.sourceLayout = ImmutableMap.copyOf(sourceLayout);
+        this.selectivityThreshold = selectivityThreshold;
+        this.isBlocked = dynamicFilter.isBlocked();
+    }
+
+    // Compiled dynamic filter is fixed per-split and generated duration page source creation.
+    // Page source implementations may subsequently implement blocking on completion of dynamic filters, but since
+    // that occurs after page source creation, we cannot be guaranteed a completed dynamic filter here for initial splits
+    public synchronized Supplier<FilterEvaluator> createDynamicPageFilterEvaluator(ColumnarFilterCompiler compiler)
+    {
+        if (isBlocked == null) {
+            return compiledDynamicFilter;
+        }
+        // Recompile dynamic filter if it has changed since last compilation
+        if (compiledDynamicFilter == null || isBlocked.isDone()) {
+            isBlocked = dynamicFilter.isBlocked();
+            boolean isAwaitable = dynamicFilter.isAwaitable();
+            TupleDomain<Symbol> currentPredicate = dynamicFilter.getCurrentPredicate().transformKeys(columnHandles::get);
+            List<Expression> expressionConjuncts = domainTranslator.toPredicateConjuncts(currentPredicate)
+                    .stream()
+                    .map(expression -> irExpressionOptimizer.process(expression, session, ImmutableMap.of()).orElse(expression))
```

Comment:
nit: I think it would be beneficial to add comment why we run it through IR optimizer here

Reply:
added comment